library/dnet.c

```diff
@@ -393,6 +404,375 @@ static int dnet_iterator_callback_send(void *priv, void *data, uint64_t dsize)
+		dnet_log(n, DNET_LOG_ERROR, "%s: Interrupting iterator because peer has been disconnected",
+				dnet_dump_id(&send->cmd.id));
+		err = -EINTR;
+		goto err_out_exit;
+	}
+
+	memset(&ctl, 0, sizeof(struct dnet_io_control));
+
+	memcpy(ctl.id.id, re->key.id, DNET_ID_SIZE);
+	ctl.id.group_id = 0; // will be rewritten in dnet_trans_create_send_all()
+
+	memcpy(ctl.io.id, re->key.id, DNET_ID_SIZE);
+	ctl.io.timestamp = re->timestamp;
+	ctl.io.user_flags = re->user_flags;
+	ctl.io.total_size = re->size;
+	ctl.io.size = re->size;
+	ctl.io.flags = DNET_IO_FLAGS_WRITE_NO_FILE_INFO;
+
+	// overwrite doesn't care whether remote key differs from local
+	// when this flag is not set, we only overwrite the same data or if there is no remote copy at all
+	if (!(send->iflags & DNET_IFLAGS_OVERWRITE))
+		ctl.io.flags |= DNET_IO_FLAGS_COMPARE_AND_SWAP;
+
+	// deliberately do not set DNET_FLAGS_NEED_ACK
+	// if WRITE command has failed, @dnet_process_cmd_with_backend_raw() will set this bit automatically,
+	// and will send acknowledge with error
+	//
+	// if there is no error, @dnet_file_info structure will be returned
+	ctl.cflags = 0;
+
+	ctl.fd = fd;
+	ctl.local_offset = data_offset;
+	ctl.cmd = DNET_CMD_WRITE;
+	ctl.ts.tv_sec = re->timestamp.tsec;
+	ctl.ts.tv_nsec = re->timestamp.tnsec;
+
+	wp = malloc(sizeof(struct dnet_iterator_server_send_write_private) + dsize);
+	if (!wp) {
+		dnet_log(n, DNET_LOG_ERROR, "%s: Interrupting iterator because failed to allocate write private data",
+				dnet_dump_id(&send->cmd.id));
+		err = -ENOMEM;
+		goto err_out_exit;
+	}
+
+	memset(wp, 0, sizeof(struct dnet_iterator_server_send_write_private));
+
+	atomic_init(&wp->refcnt, send->group_num);
+	wp->send = send;
+
+	memcpy(wp->data, re, dsize);
+	wp->dsize = dsize;
+
+	ctl.complete = dnet_iterator_server_send_complete;
+	ctl.priv = wp;
+
+	s = dnet_session_create(n);
+	if (!s) {
+		dnet_log(n, DNET_LOG_ERROR, "%s: Interrupting iterator because failed to create new session",
+				dnet_dump_id(&send->cmd.id));
+		err = -ENOMEM;
+		goto err_out_free;
+	}
+
+	dnet_session_set_trace_id(s, send->cmd.trace_id);
+
+	err = dnet_session_set_groups(s, send->groups, send->group_num);
+	if (err) {
+		dnet_log(n, DNET_LOG_ERROR, "%s: Interrupting iterator because failed to set %d groups",
+				dnet_dump_id(&send->cmd.id), send->group_num);
+		err = -ENOMEM;
+		goto err_out_session_destroy;
+	}
+
+	atomic_add(&send->bytes_pending, re->size);
+
+	dnet_log(n, DNET_LOG_NOTICE, "%s: %s: sending WRITE request, iterator response: %s, user_flags: %llx, ts: %lld.%09lld, "
+			"status: %d, size: %lld, iterated_keys: %lld/%lld",
+			__func__,
+			dnet_dump_id(&send->cmd.id), dnet_dump_id_str(re->key.id),
+			(unsigned long long)re->user_flags,
+			(unsigned long long)re->timestamp.tsec, (unsigned long long)re->timestamp.tnsec,
+			re->status, (unsigned long long)re->size,
+			(unsigned long long)re->iterated_keys, (unsigned long long)re->total_keys);
+	dnet_convert_iterator_response(re);
```

Comment:
I've reworked this chunk a bit to eliminate excessive conversions, it is not needed